File: app/models/place/finish_line.rb
class Place::FinishLine < ApplicationRecord
  belongs_to :place
  has_many :virtual_competitions, dependent: :restrict_with_error

  validates :name, presence: true
  validates :start_latitude, :end_latitude,
            numericality: { greater_than_or_equal_to: -90, less_than_or_equal_to: 90 }

  validates :start_longitude, :end_longitude,
            numericality: { greater_than_or_equal_to: -180, less_than_or_equal_to: 180 }

  def to_coordinates
    [
      { latitude: start_latitude, longitude: start_longitude },
      { latitude: end_latitude,   longitude: end_longitude   }
    ]
  end

  def center
    latitude = start_latitude + ((end_latitude - start_latitude) / 2)
    longitude = start_longitude + ((end_longitude - start_longitude) / 2)
    {
      latitude: latitude,
      longitude: longitude
    }
  end
end

File: app/javascript/utils/tournaments/map/finish_line.js
export default class FinishLine {
  constructor(points) {
    this.points = points
  }

  get path() {
    return this.points.map(el => {
      return {
        lat: Number(el.latitude),
        lng: Number(el.longitude)
      }
    })
  }

  get center() {
    return {
      lat: this.start_latitude + (this.end_latitude - this.start_latitude) / 2,
      lng: this.start_longitude + (this.end_longitude - this.start_longitude) / 2
    }
  }

  get start_latitude() {
    return Number(this.start_point.latitude)
  }

  get start_longitude() {
    return Number(this.start_point.longitude)
  }

  get end_latitude() {
    return Number(this.end_point.latitude)
  }

  get end_longitude() {
    return Number(this.end_point.longitude)
  }

  get start_point() {
    return this.points[0]
  }

  get end_point() {
    return this.points[1]
  }
}
